### PR TITLE
make pod with extended resources pass predicate check when kubelet restart

### PR DESCRIPTION
**Pull request by ddongchen**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
/kind bug

**What this PR does / why we need it**:
   when kubelet restart, all extended resources will be zero. And during Predicate,
the plugin extended resources will be recoverd to pass the general check,
while other extended resources is still zero. This result all pods that has already requested
the extended resources to fail.

**Which issue(s) this PR fixes**:

Fixes #  https://github.com/kubernetes/kubernetes/issues/73948 and https://github.com/kubernetes/kubernetes/issues/64632, 
   For there is already extended resources in use for some production environment, we should ensure pods with extended resources not fail when kubelet restart

**Special notes for your reviewer**:
@jiayingz 
**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/kubelet/lifecycle/predicate.go [hunk 1]
```diff
@@ -172,3 +175,50 @@ func removeMissingExtendedResources(pod *v1.Pod, nodeInfo *schedulernodeinfo.Nod
...
+// the node to pass GeneralPredicates admission checking.
+func setNodeAllocatableScalarResources(nodeInfo *schedulercache.NodeInfo, admitPod *v1.Pod) {
+	var newAllocatableResource *schedulercache.Resource
+	var needUpdate bool = false
+	allocatableResource := nodeInfo.AllocatableResource()
+	if allocatableResource.ScalarResources == nil {
+		allocatableResource.ScalarResources = make(map[v1.ResourceName]int64)
+	}
+
+	ScalarResources := predicates.GetResourceRequest(admitPod).ScalarResources
```
Comment (mattjmcnaughton) [inline pkg/kubelet/lifecycle/predicate.go]: Nit: Can you share more about why we deviated from the `camelCase` naming convention here?

**Review by mattjmcnaughton — COMMENTED** [review]
Thanks for your pr! I'll let someone with more context on this part of the architecture do a larger review, but just some small stylistic nits from me.

Also, can you share more about your decision not to add unit tests? Again, I'm not super familiar with this part of the system so just curious :) It seems like, if this code isn't already covered by existing unit tests, and its not unacceptably difficult to test, it'd be good to have unit test coverage for this addition.

**Comment by jiayingz** [comment]
Thanks a lot for looping me in, @ddongchen 

Overall, I feel this is another hacky workaround for the extended resource life cycle problem we have discussed multiple times in the past. I can totally understand why you need to make this change. However, my feeling is if we don't come up with a long term solution for the underlying problem, the workaround we put in here may easily cause issues in the future.

I am hoping we can spend some time to discuss some high level questions first:

- Do we want to support the use case of directly patching node API object to export an extended resource? From security and maintenance points, should we instead require that node resource exporting has to go through a local agent?

- What kind of resource admission guarantees we can provide for extended resource that is exported dynamically from Kubelet life cycle, given the facts that node local state can be gone due to node recreation, node API object can be deleted and recreated when node is unavailable long enough, there is no guaranteed ordering between pod restart and Kubelet re-syncing with API server, and Kubelet resource admission is supposed to serve as last defense, e.g., in the presence of multiple schedulers.

I am adding more folks for these general discussions.
/cc @dchen1107 @derekwaynecarr @ConnorDoyle @vishh  @liggitt @thockin

File: pkg/kubelet/lifecycle/predicate.go (hunk 1 above)
Comment (ddongchen) [reply pkg/kubelet/lifecycle/predicate.go]: so sorry, it's my fault, it should be "scalarResources", i will fix it immediately

**Review by ddongchen — COMMENTED** [review]
> Thanks for your pr! I'll let someone with more context on this part of the architecture do a larger review, but just some small stylistic nits from me.
> 
> Also, can you share more about your decision not to add unit tests? Again, I'm not super familiar with this part of the system so just curious :) It seems like, if this code isn't already covered by existing unit tests, and its not unacceptably difficult to test, it'd be good to have unit test coverage for this addition.

I membered I have added "I have not add the unit test case, we can just have a talk".  so sorry for this. For this commit, we can first have a talk about if this is a good  option.

**Comment by ddongchen** [comment]
> Thanks a lot for looping me in, @ddongchen
> 
> Overall, I feel this is another hacky workaround for the extended resource life cycle problem we have discussed multiple times in the past. I can totally understand why you need to make this change. However, my feeling is if we don't come up with a long term solution for the underlying problem, the workaround we put in here may easily cause issues in the future.
> 
> I am hoping we can spend some time to discuss some high level questions first:
> 
> * Do we want to support the use case of directly patching node API object to export an extended resource? From security and maintenance points, should we instead require that node resource exporting has to go through a local agent?
> * What kind of resource admission guarantees we can provide for extended resource that is exported dynamically from Kubelet life cycle, given the facts that node local state can be gone due to node recreation, node API object can be deleted and recreated when node is unavailable long enough, there is no guaranteed ordering between pod restart and Kubelet re-syncing with API server, and Kubelet resource admission is supposed to serve as last defense, e.g., in the presence of multiple schedulers.
> 
> I am adding more folks for these general discussions.
> /cc @dchen1107 @derekwaynecarr @ConnorDoyle @vishh @liggitt @thockin

hi,

> Thanks a lot for looping me in, @ddongchen
> 
> Overall, I feel this is another hacky workaround for the extended resource life cycle problem we have discussed multiple times in the past. I can totally understand why you need to make this change. However, my feeling is if we don't come up with a long term solution for the underlying problem, the workaround we put in here may easily cause issues in the future.
> 
> I am hoping we can spend some time to discuss some high level questions first:
> 
> * Do we want to support the use case of directly patching node API object to export an extended resource? From security and maintenance points, should we instead require that node resource exporting has to go through a local agent?
> * What kind of resource admission guarantees we can provide for extended resource that is exported dynamically from Kubelet life cycle, given the facts that node local state can be gone due to node recreation, node API object can be deleted and recreated when node is unavailable long enough, there is no guaranteed ordering between pod restart and Kubelet re-syncing with API server, and Kubelet resource admission is supposed to serve as last defense, e.g., in the presence of multiple schedulers.
> 
> I am adding more folks for these general discussions.
> /cc @dchen1107 @derekwaynecarr @ConnorDoyle @vishh @liggitt @thockin

hi, 

> Thanks a lot for looping me in, @ddongchen
> 
> Overall, I feel this is another hacky workaround for the extended resource life cycle problem we have discussed multiple times in the past. I can totally understand why you need to make this change. However, my feeling is if we don't come up with a long term solution for the underlying problem, the workaround we put in here may easily cause issues in the future.
> 
> I am hoping we can spend some time to discuss some high level questions first:
> 
> * Do we want to support the use case of directly patching node API object to export an extended resource? From security and maintenance points, should we instead require that node resource exporting has to go through a local agent?
> * What kind of resource admission guarantees we can provide for extended resource that is exported dynamically from Kubelet life cycle, given the facts that node local state can be gone due to node recreation, node API object can be deleted and recreated when node is unavailable long enough, there is no guaranteed ordering between pod restart and Kubelet re-syncing with API server, and Kubelet resource admission is supposed to serve as last defense, e.g., in the presence of multiple schedulers.
> 
> I am adding more folks for these general discussions.
> /cc @dchen1107 @derekwaynecarr @ConnorDoyle @vishh @liggitt @thockin

hi,jiayingz. I know there may need a more established management for the extended resources. while I think we should ensure pod not restart when update to higher version, especially in production environment.  
The issue is from https://github.com/kubernetes/kubernetes/pull/64784, if
the above patch is not very fit, shall we delete the extended resource, instead of zero, like this:

func (kl *Kubelet) reconcileExtendedResource(initialNode, node *v1.Node) bool {
-                       node.Status.Capacity[k] = *resource.NewQuantity(int64(0), resource.DecimalSI)
-                       node.Status.Allocatable[k] = *resource.NewQuantity(int64(0), resource.DecimalSI)
+                       delete(node.Status.Capacity, k)
+                       delete(node.Status.Allocatable, k)

That is removing extended resources when kubelet register to api server.  For device plugin, this work well, and for other extended resources is also well, because during predicate, it will first remove missing extended resources:

podWithoutMissingExtendedResources := removeMissingExtendedResources(admitPod, nodeInfo)

is this ok?

**Comment by ddongchen** [comment]
> Thanks a lot for looping me in, @ddongchen
> 
> Overall, I feel this is another hacky workaround for the extended resource life cycle problem we have discussed multiple times in the past. I can totally understand why you need to make this change. However, my feeling is if we don't come up with a long term solution for the underlying problem, the workaround we put in here may easily cause issues in the future.
> 
> I am hoping we can spend some time to discuss some high level questions first:
> 
> * Do we want to support the use case of directly patching node API object to export an extended resource? From security and maintenance points, should we instead require that node resource exporting has to go through a local agent?
> * What kind of resource admission guarantees we can provide for extended resource that is exported dynamically from Kubelet life cycle, given the facts that node local state can be gone due to node recreation, node API object can be deleted and recreated when node is unavailable long enough, there is no guaranteed ordering between pod restart and Kubelet re-syncing with API server, and Kubelet resource admission is supposed to serve as last defense, e.g., in the presence of multiple schedulers.
> 
> I am adding more folks for these general discussions.
> /cc @dchen1107 @derekwaynecarr @ConnorDoyle @vishh @liggitt @thockin

hi,jiayingz. I know there may need a more established management for the extended resources. while I think we should ensure pod not restart when update to higher version, especially in production environment.  
The issue is from https://github.com/kubernetes/kubernetes/pull/64784, if
the above patch is not very fit, shall we delete the extended resource, instead of zero, like this:
image

That is removing extended resources when kubelet register to api server.  For device plugin, this work well, and for other extended resources is also well, because during predicate, it will first remove missing extended resources:

podWithoutMissingExtendedResources := removeMissingExtendedResources(admitPod, nodeInfo)

is this ok?

**Comment by jiayingz** [comment]
@ddongchen if let me vote, I would rather we roll back #64784 instead of adding another hack on top of it.

I also feel we should correct k8s extended-resource doc to make sure the feature state matches its actual quality. The device plugin feature is currently in beta. We do have test coverage from e2e_node, e2e, cluster/master upgrade that exercises its life cycle with Kubelet and node life cycles. However, I don't think we can graduate it to GA until we come up with a solid long-term solution for its life cycle issues. In comparison, exporting extended resource by directly patching node object doesn't have any e2e test coverage. I feel it should really be marked as at alpha state to make it clear that people may likely run into issues when using it on production.

I know this may be disturbing for people who are already using this feature in production. I hope we can work together to come up a plan that we can really graduate the feature to GA.

**Comment by ddongchen** [comment]
> @ddongchen if let me vote, I would rather we roll back #64784 instead of adding another hack on top of it.
> 
> I also feel we should correct k8s extended-resource doc to make sure the feature state matches its actual quality. The device plugin feature is currently in beta. We do have test coverage from e2e_node, e2e, cluster/master upgrade that exercises its life cycle with Kubelet and node life cycles. However, I don't think we can graduate it to GA until we come up with a solid long-term solution for its life cycle issues. In comparison, exporting extended resource by directly patching node object doesn't have any e2e test coverage. I feel it should really be marked as at alpha state to make it clear that people may likely run into issues when using it on production.
> 
> I know this may be disturbing for people who are already using this feature in production. I hope we can work together to come up a plan that we can really graduate the feature to GA.

so, shall we roll back https://github.com/kubernetes/kubernetes/pull/64784 until there is an plan? Otherwise, I'm worried about other people will still use extended resources in their production, while not notice this action.

**Comment by chenchun** [comment]
@jiayingz is there a final result of this? does #64784 going to be rollbacked or should we write a device plugin in order to make pods with extended resource to survive Kubelet restart ?

**Comment by jiayingz** [comment]
@chenchun @ddongchen Thinking about this a bit more, in 1.15 time frame, instead of rolling back  #64784 entirely, how about we change it to calling an interface function exposed by cm.devicemanager, say ResetExtendedResourceCapacity, that:
- returns false if DevicePlugin feature is disabled
- returns false if a local checkpoint file exists (note we need to extend cm.devicemanager to always generates a local checkpoint file when it starts)
- returns true if DevicePlugin feature is enabled and no checkpoint exists -- we can use this as a strong indicator that node instance has been recreated
Then we can modify reconcileExtendedResource() to only zero out extended resource capacity if ResetExtendedResourceCapacity returns true?

Longer term, I still feel directly patching node status capacity/allocatable fields without going through kubelet is perhaps problematic. E.g., node API object can be deleted in certain cases, and it would be hard to ensure the guarantee people might expect (e.g., if I set capacity for a given resource to be two, then it is guaranteed pods can't consume more than two resources) when node API object is recreated with the current open admission policy.

**Comment by chenchun** [comment]
I agree the way of calling devicemanager if the resource is managed by itself and only zero out those that needs to be done. I never tried device plugin before, so I may not understand the details of `ResetExtendedResourceCapacity` in your comment.

> when node API object is recreated with the current open admission policy

I searched the usage of corev1.Nodes().Delete() and it seems https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_lifecycle_controller.go#L193 is the only place that invoked this func. If I understand correctly, this would happen only on cloud provider. I'm not running kubernetes on a cloud provider, so node object wouldn't be deleted unless someone deletes it with kubectl?

I wrote a central controller to patch node status capacity/allocatable fields and this is a simpler way for me instead of device plugin running on each node.

**Comment by ddongchen** [comment]
> @chenchun @ddongchen Thinking about this a bit more, in 1.15 time frame, instead of rolling back #64784 entirely, how about we change it to calling an interface function exposed by cm.devicemanager, say ResetExtendedResourceCapacity, that:

@jiayingz do you mean the extended resources, which added with patch, should be managed by device plugin manager? If yes, for "returns true if DevicePlugin feature is enabled and no checkpoint exists", is this mean just the checkpoint file existed ? or the related extended resource should be in the checkpoint file? if is the second, then we should save extended resource to the checkpoint file.

**Comment by chardch** [comment]
I'll help out and work on this issue to check with the device manager in order to determine whether or not to zero out the extended resources.

**Comment by chardch** [comment]
> @chenchun @ddongchen Thinking about this a bit more, in 1.15 time frame, instead of rolling back #64784 entirely, how about we change it to calling an interface function exposed by cm.devicemanager, say ResetExtendedResourceCapacity, that:
> 
> * returns false if DevicePlugin feature is disabled
> * returns false if a local checkpoint file exists (note we need to extend cm.devicemanager to always generates a local checkpoint file when it starts)
> * returns true if DevicePlugin feature is enabled and no checkpoint exists -- we can use this as a strong indicator that node instance has been recreated
>   Then we can modify reconcileExtendedResource() to only zero out extended resource capacity if ResetExtendedResourceCapacity returns true?

#77699 is the PR to address the above comment.

**Comment by chardch** [comment]
> * returns false if a local checkpoint file exists (note we need to extend cm.devicemanager to always generates a local checkpoint file when it starts)
> * returns true if DevicePlugin feature is enabled and no checkpoint exists -- we can use this as a strong indicator that node instance has been recreated

If the local checkpoint is written when the devicemanager starts, then wouldn't this method always return false? Since even on a node recreation, once the devicemanager is started the checkpoint will be written again before we could call ShouldResetExtendedResourceCapacity.

**Comment by derekwaynecarr** [comment]
i tend to agree with @jiayingz that we should enumerate the known gaps around device plugins as they exist today, and agree on a path forward before adding more incremental patches that may not be obvious.  

can we write up a node_e2e test that demonstrates what is broken first?

**Pull request closed without merging**